core/commands/diff.py

```diff
@@ -395,10 +453,37 @@ class GsDiffNavigateCommand(GsNavigate):
 
     offset = 0
 
-    def run(self, edit, **kwargs):
-        super().run(edit, **kwargs)
-        self.view.run_command("show_at_center")
-
     def get_available_regions(self):
         return [self.view.line(region) for region in
                 self.view.find_by_selector("meta.diff.range.unified")]
+
+
+class GsDiffUndo(TextCommand, GitCommand):
+
+    """
+    Undo the last action taken in the diff view, if possible.
+    """
+
+    def run(self, edit):
+        sublime.set_timeout_async(self.run_async)
+
+    def run_async(self):
+        history = self.view.settings().get("git_savvy.diff_view.history") or []
+        if not history:
+            window = self.view.window()
+            if window:
+                window.status_message("Undo stack is empty")
+            return
+
+        args, stdin, cursor = history.pop()
+        # Toggle the `--reverse` flag.
+        args[1] = "-R" if not args[1] else None
+
+        self.git(*args, stdin=stdin)
+        self.view.settings().set("git_savvy.diff_view.history", history)
+        self.view.settings().set("git_savvy.diff_view.just_hunked", stdin)
+
+        self.view.run_command("gs_diff_refresh")
+        self.view.sel().clear()
+        self.view.sel().add(cursor)
+        self.view.show(cursor)
```

Comment:
FIXME: `cursor` is only valid if the user is still on the same staging area, t.i. did not '<tab>'.